Core/src/Material/SurfaceMaterialMapper.cpp

```diff
@@ -279,12 +287,71 @@ void Acts::SurfaceMaterialMapper::mapMaterialTrack(
       ++rmIter;
       continue;
     }
-    if (sfIter != mappingSurfaces.end() - 1 &&
-        (rmIter->position - sfIter->position).norm() >
-            (rmIter->position - (sfIter + 1)->position).norm()) {
-      // Switch to next assignment surface
-      ++sfIter;
+    // Do we need to switch to next assignment surface ?
+    if (sfIter != mappingSurfaces.end() - 1) {
+      int mappingType = sfIter->surface->surfaceMaterial()->mappingType();
+      int nextMappingType =
+          (sfIter + 1)->surface->surfaceMaterial()->mappingType();
+
+      if (mappingType == Acts::MappingType::PreMapping ||
+          mappingType == Acts::MappingType::Sensor) {
+        // Change surface if the material after the current surface.
+        if ((rmIter->position - mTrack.first.first).norm() >
+            (sfIter->position - mTrack.first.first).norm()) {
+          if (nextMappingType == Acts::MappingType::PostMapping) {
+            ACTS_WARNING(
+                "PreMapping or Sensor surface followed by PostMapping. Some "
+                "material "
+                "from before the PostMapping surface will be mapped onto it");
+          }
+          ++sfIter;
+        }
+      } else if (mappingType == Acts::MappingType::Default ||
+                 mappingType == Acts::MappingType::PostMapping) {
+        switch (nextMappingType) {
+          case Acts::MappingType::PreMapping:
+          case Acts::MappingType::Default: {
+            // Change surface if the material closest to the next surface.
+            if ((rmIter->position - sfIter->position).norm() >
+                (rmIter->position - (sfIter + 1)->position).norm()) {
+              ++sfIter;
+            }
+            break;
+          }
+          case Acts::MappingType::PostMapping: {
+            // Change surface if the material after the next surface.
+            if ((rmIter->position - sfIter->position).norm() >
+                ((sfIter + 1)->position - sfIter->position).norm()) {
+              ++sfIter;
+            }
+            break;
+          }
+          case Acts::MappingType::Sensor: {
+            // Change surface if the next material after the next surface.
+            if ((rmIter == rMaterial.end() - 1) ||
+                ((rmIter + 1)->position - sfIter->position).norm() >
+                    ((sfIter + 1)->position - sfIter->position).norm()) {
+              ++sfIter;
+            }
+            break;
+          }
+          default: {
+            ACTS_ERROR("Incorect mapping type for the next surface : "
+                       << (sfIter + 1)->surface->geometryId());
+          }
+        }
+      } else {
+        ACTS_ERROR("Incorect mapping type for surface : "
+                   << sfIter->surface->geometryId());
+      }
     }
+
```

Comment:
What's with the comment block?

Reply:
That was the previous way of handeling the surface association that I forgot to remove... Thanks for catching that.